File: ChemistryBalancing/balance.js
const equationHolder = document.getElementById('equationHolder');
const balanceButton = document.getElementById('balanceButton');
const answerText = document.getElementById('answer');

async function waitTick() {
    return new Promise((results) => {
        setTimeout(() => {
            results();
        }, 0.0025);
    });
};

async function balance(source) {
    class equationClass {
        reactants = [];
        products = [];
    };
    function makeParts(equationSection, part) {
        const sections = part.split('+').length;
        let startIndex = 0;
        for (let i = 0; i < sections; i++) {
            const endIndex = (part.indexOf('+') != -1 ? part.indexOf('+') : part.length);
            const portion = part.slice(startIndex, endIndex);
            const portionSize = portion.length;
            const portionTable = {}
            let element = '';
            let number = '';
            console.log(portionSize);
            for (let j = 0; j < portionSize; j++) {
                const character = portion[j];
                if (isNaN(character)) {
                    if (number) {
                        portionTable[element] = number;
                        element = '';
                        number = ''
                    };
                    element += character;
                } else {
                    number += character;
                };
            };
            if (number) {
                portionTable[element] = number;
                element = '';
                number = ''
            };
            equationSection.push(portionTable);
            part = part.replace('+', '');
            startIndex = endIndex;
        };
        return(sections);
    };

    function multiplyAndCalculate(equationTable, multiplier) {
        function multiply(section, sectionLength, multiplierStart) {
            for (let i = 0; i < sectionLength; i++) {
                const portion = section[i];
                const portionLength = Object.keys(portion).length;
                for (let j = 0; j < portionLength; j++) {
                    const element = Object.keys(portion)[j];
                    portion[element] *= multiplier[i+multiplierStart];
                };
            };
        };

        function merge(section, sectionLength) {
            for (let i = 1; i < sectionLength; i++) {
                const portion = section[i];
                const portionLength = Object.keys(portion).length;
                for (let j = 0; j < portionLength; j++) {
                    const element = Object.keys(portion)[j];
                    if (element in section[0]) {
                        section[0][element] += portion[element];
                    } else {
                        section[0][element] = portion[element];
                    };
                };
            };
            section.splice(1, sectionLength);
        };

        const reactantsLength = equationTable.reactants.length;
        const productsLength = equationTable.products.length;
        multiply(equationTable.reactants, reactantsLength, 0);
        multiply(equationTable.products, productsLength, reactantsLength);
        merge(equationTable.reactants, reactantsLength);
        merge(equationTable.products, productsLength);
        const allReactantsLength = Object.keys(equationTable.reactants[0]).length;
        for (let i = 0; i < allReactantsLength; i++) {
            const element = Object.keys(equationTable.reactants[0])[i]
            if (element in equationTable.products[0]) {
                if (equationTable.reactants[0][element] != equationTable.products[0][element]) {
                    return(false);
                };
            } else {
                answerText.textContent = 'Unclosed Equation!';
                throw new Error('Unclosed Equation!');
            };
        };
        return(true);
    };

    function nextMultiplier(index, length, reverse) {
        const combination = [];
        for (let i = 0; i < length; i++) {
            const digit = (!reverse ? (index%9)+1 : 9 - (index%9));
            combination.unshift(digit);
            index = Math.floor(index/9);
        };
        return(combination);
    };
    let equation = source.replaceAll(' ','');
    const midPoint = equation.indexOf('>');
    if (midPoint <= 0) {
        answerText.textContent = 'Invalid Equation!';
        throw new Error('Invalid Equation!');
    };
    let reactants = equation.slice(0, midPoint);
    let products = equation.slice(midPoint+1, equation.length);
    const equationTable = new equationClass;
    let totalParts = 0;
    totalParts += makeParts(equationTable.reactants, reactants, true);
    totalParts += makeParts(equationTable.products, products, false);
    let multiplier = [];
    const multiplierLength = equationTable.reactants.length + equationTable.products.length;
    for (let i = 0; i < multiplierLength; i++) {
        multiplier.push(1);
    };
    let multiplierI = 1;
    const maxLimit = (9**totalParts)/2;

    function tryMultiplier(reverse) {
        multiplier = nextMultiplier(multiplierI, multiplierLength, reverse);
        const solvedEquation = multiplyAndCalculate(structuredClone(equationTable), multiplier);
        return(solvedEquation);
    };

    while (true) {;
        answerText.textContent = `Max wait time: ~${(Math.round((maxLimit-multiplierI)*0.0000125)*1000)/1000}sec`
        const solvedEquation = tryMultiplier(false);
        if (solvedEquation) {
            return(`Multipliers are: ${multiplier}`);
        };

        const solvedReverseEquation = tryMultiplier(true);
        if (solvedReverseEquation) {
            return(`Multipliers are: ${multiplier}`);
        };

        multiplierI += 1;
        if (multiplierI >= maxLimit) {
            answerText.textContent = 'Equation Unsolvable/Too Big!';
            throw new Error('Equation Unsolvable/Too Big!');
        };

        if (!(multiplierI % 1000)) {
            await waitTick();
        };
    };
};

//Example Problem: Rb3P1O4 + H2O1 > Rb1O1H1 + H3P1O4 
balanceButton.addEventListener('click', async function() {
    answerText.textContent = 'Multiplier: Waiting!'
    const multiplier =  await balance(equationHolder.value);
    answerText.textContent = multiplier;
});

File: cIntroCode/balance.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <stdexcept>
#include <map>
#include <vector>
#include <ctype.h>
#include <thread>
#include <chrono>
#include <cmath>

class equationClass {
    public:
        std::vector<std::map<std::string, int>> reactants;
        std::vector<std::map<std::string, int>> products;
};


int makeParts(std::vector<std::map<std::string, int>>& equationSection, std::string part) {
    const int sections = std::count(part.begin(), part.end(), '+')+1;
    int startIndex = 0;
    
    for (int i = 0; i < sections; i++) {
        const size_t plusPos = part.find('+', startIndex);
        const int endIndex = (plusPos != std::string::npos) ? plusPos : part.length();
        const std::string portion = part.substr(startIndex, endIndex - startIndex);
        const int portionSize = portion.length();
        std::cout << "Portion Size: " << portionSize << std::endl;
        std::map<std::string, int> portionMap;
        std::string element;
        std::string number; 
        for (int j = 0; j < portionSize; j++) {
            const char character = portion[j];
            std::cout << "Character: " << character << std::endl;
            if (!isdigit(character)) {
                if (number.length() > 0) {
                    portionMap[element] = std::stoi(number);
                    element = "";
                    number = "";
                };
                element += character;
            } else {
                number += character;
            };
        };
        if (number.length() > 0) {
            portionMap[element] = std::stoi(number);
        };

        equationSection.push_back(portionMap);
        if (plusPos != std::string::npos) {
            startIndex = plusPos + 1;
        } else {
            break;
        };
    };

    return sections;
};

std::vector<int> nextMultiplier(int index, int length, bool reverse) {
    std::vector<int> combination;
    for (int i = 0; i < length; i++) {
        const int digit = (!reverse ? (index%9)+1 : 9 - (index%9));
        combination.insert(combination.begin(), digit);
        index = std::floor(index/9);
    };
    return combination;
};

bool multiplyAndCalculate(equationClass equationTable, int multiplierI, std::vector<int> multiplier) {
    auto multiply = [multiplier] (std::vector<std::map<std::string, int>>& section, int sectionLength, int multiplierStart) {
        for (int i = 0; i < sectionLength; i++) {
            std::map<std::string, int>& portion = section[i];
            for (auto& pair : portion) {
                pair.second *= multiplier[i+multiplierStart];
            };
        };
    };

    auto merge = [] (std::vector<std::map<std::string, int>>& section, int sectionLength) {
        for (int i = 1; i < sectionLength; i++) {
            std::map<std::string, int>& portion = section[i];
            for (auto& pair : portion) {
                if (section[0].find(pair.first) != section[0].end()) {
                    section[0][pair.first] += pair.second;
                } else {
                    section[0][pair.first] = pair.second;
                };
            };
        };
        section.erase(section.begin()+1, section.end());
    };

    const int reactantsLength = equationTable.reactants.size();
    const int productsLength = equationTable.products.size(); 

    /*//Debug
    std::cout << "Multiplier: ";
    for (int i = 0; i < multiplier.size(); i++) {
        std::cout << multiplier[i];
    };
    std::cout << std::endl;*/

    multiply(equationTable.reactants, reactantsLength, 0);
    multiply(equationTable.products, productsLength, reactantsLength);

    merge(equationTable.reactants, reactantsLength);
    merge(equationTable.products, productsLength);

    for (auto pair = equationTable.reactants[0].begin(); pair != equationTable.reactants[0].end(); pair++) {
        if (equationTable.products[0].find(pair->first) != equationTable.products[0].end()) {
            if (equationTable.products[0][pair->first] != equationTable.reactants[0][pair->first]) {
                return false;
            };
        } else {
            throw std::invalid_argument("Unclosed Equation!");
        };
    };

    return true;
};

std::string balanceEquation(std::string source) {
    source.erase(std::remove(source.begin(), source.end(), ' '), source.end());
    int midPoint = source.find('>', 0);
    if (midPoint <= 0) {
        throw std::invalid_argument("Invalid Equation!");
    };

    std::string reactants = source.substr(0, midPoint);
    std::string products = source.substr(midPoint+1, source.length());

    equationClass equationTable;
    int totalParts = 0;
    totalParts += makeParts(equationTable.reactants, reactants);
    totalParts += makeParts(equationTable.products, products);
    std::vector<int> multiplier;
    const int multiplierLength = (equationTable.reactants.size() + equationTable.products.size());
    for (int i = 0; i < multiplierLength; i++) {
        multiplier.push_back(1);
    };
    int multiplierI = 1;
    const std::int64_t maxLimit = (std::pow(9, totalParts))/2;

    /*//Debug
    for (int i = 0; i < equationTable.reactants.size(); i++) {
        std::map<std::string, int>& portion = equationTable.reactants[i];
        std::cout << "Pair: ";
        for (auto& pair : portion) {
            std::cout << pair.first << ":" << pair.second << ", ";
        };
        std::cout << "---------------";
    };
    std::cout << std::endl << "Products VVV" << std::endl;
    for (int i = 0; i < equationTable.products.size(); i++) {
        std::map<std::string, int>& portion = equationTable.products[i];
        std::cout << "Pair: ";
        for (auto& pair : portion) {
            std::cout << pair.first << ":" << pair.second << ", ";
        };
        std::cout << "---------------";
    };
    std::cout << std::endl << std::endl;*/

    auto tryMultiplier = [&multiplierI, multiplierLength, &multiplier, equationTable](bool reverse) {
        multiplier = nextMultiplier(multiplierI, multiplierLength, reverse);
        const bool solvedEquation = multiplyAndCalculate(equationTable, multiplierI, multiplier);
        return solvedEquation;
    };

    while (true) {
        if (tryMultiplier(false)) {
            std::cout << "Multiplier: ";
            for (int i = 0; i < multiplierLength; i++) {
                std::cout << multiplier[i];
            };
            std::cout << std::endl;
            return "Convert multiplier vector to string!";
        };

        if (tryMultiplier(true)) {
            std::cout << "Multiplier: ";
            for (int i = 0; i < multiplierLength; i++) {
                std::cout << multiplier[i];
            };
            std::cout << std::endl;
            return "Convert multiplier vector to string!";
        };

        multiplierI += 1;
        if (multiplierI >= maxLimit) {
            throw std::invalid_argument("Equation Unsolvable/Too Big!");
        };

        if (!(multiplierI % 50000)) {
            std::this_thread::sleep_for(std::chrono::milliseconds(1));
        };
    };

    return "something went wrong!";
}

int main() {
    std::string balancedEquation = balanceEquation("A1 + B2 + C3 + D2 + E1 > E9 + D8 + C7 + B8 + A9");
    std::cout << balancedEquation;
};